resource-aggregate: add update device metadata commands

jkralik opened the pull request.

jkralik committed:
WIP

ondrejtomcik commented on resource-aggregate/pb/events.proto:
```diff
@@ -113,4 +113,12 @@ message ResourceStateSnapshotTaken {
+
+message DeviceCloudStatusSnapshotTaken {
+    string device_id = 1;
+    OnlineState online = 2;
```
Status status = 2;

jkralik replied:
We have `message Status` so it will be `message OnlineStatus`

jkralik committed:
add DeviceCloudStatusUpdated

ondrejtomcik commented on resource-aggregate/pb/events.proto:
```diff
@@ -113,4 +113,20 @@ message ResourceStateSnapshotTaken {
+    EventMetadata event_metadata = 101;
+}
+
+message DeviceCloudStatusSnapshotTaken {
```
DeviceMetadataSnapshotTaken

ondrejtomcik commented on resource-aggregate/pb/events.proto:
```diff
@@ -113,4 +113,20 @@ message ResourceStateSnapshotTaken {
     EventMetadata event_metadata = 101;
+}
+
+message DeviceCloudStatusUpdated {
```
DeviceMetadataUpdated
But do we need it? Don't we want to have it more granular - more events, e.g. DeviceStatusChanged, ...? gRPC Gateway API will provide definitely more granular events.

ondrejtomcik commented on resource-aggregate/pb/events.proto:
```diff
@@ -113,4 +113,20 @@ message ResourceStateSnapshotTaken {
+
+message DeviceCloudStatusUpdated {
+    string device_id = 1;
+    OnlineState online = 2;
```
just `Status`

ondrejtomcik commented on resource-aggregate/pb/service.proto:
```diff
@@ -18,4 +18,5 @@ service ResourceAggregate {
   rpc ConfirmResourceDelete (ConfirmResourceDeleteRequest) returns (ConfirmResourceDeleteResponse) {}
   rpc CreateResource (CreateResourceRequest) returns (CreateResourceResponse) {}
   rpc ConfirmResourceCreate (ConfirmResourceCreateRequest) returns (ConfirmResourceCreateResponse) {}
+  rpc UpdateDeviceCloudStatus (UpdateDeviceCloudStatusRequest) returns (UpdateDeviceCloudStatusResponse) {}
```
UpdateDeviceMetadata

ondrejtomcik commented on resource-aggregate/pb/events.proto:
```diff
@@ -113,4 +113,20 @@ message ResourceStateSnapshotTaken {
+message DeviceCloudStatusSnapshotTaken {
+    string device_id = 1;
+    OnlineState online = 2;
+    ResourceObservationsState resource_observations = 3;
```
Observation is CoAP specific term. By this name, you describe tech solution hidden behind device/resource shadow. The user is not disabling observations, he doesn't have to know what it is.
`ShadowSync`:`Enabled`/`Disables`

jkralik committed:
update protos according suggestions

jkralik committed:
use oneof in command and event

ondrejtomcik commented on resource-aggregate/pb/commands.proto:
```diff
@@ -385,3 +385,24 @@ message ConfirmResourceCreateResponse {
+message UpdateDeviceMetadataRequest {
+    string device_id = 1;
+    oneof update {
+        OnlineStatus online = 2;
```
ConnectionStatus status = 2;

ondrejtomcik commented on resource-aggregate/pb/commands.proto:
```diff
@@ -385,3 +385,24 @@ message ConfirmResourceCreateResponse {
 }
 
+message OnlineStatus {
+    bool value = 1;
```
I would change it to enum (string). online/offline for now. We might extend it easily by connected, authorized, online ...

jkralik committed:
add deviceMetadata events and grpc

jkralik committed:
resource-aggregate: add update device metadata commands

- allow to disable resource shadow at the device
- allow to update device connection status

ondrejtomcik commented on resource-aggregate/commands/utils.go:
```diff
@@ -70,3 +71,14 @@ func HTTPStatus2Status(s int) Status {
+	if s.Value == ConnectionStatus_OFFLINE {
+		return false
+	}
+	if s.ValidUntil <= 0 {
```
When it can be <= 0? could you please write it there?

jkralik replied:
fixed

ondrejtomcik commented on resource-aggregate/events/deviceMetadataSnapshotTaken.go:
```diff
@@ -0,0 +1,219 @@
+
+		switch {
+		case req.GetStatus() != nil:
+			// it is expected that the device directly update self status, so we don't need confirm it.
```
`it is expected that the device updates the status on its own. no confirmation needed.`

jkralik replied:
fixed

ondrejtomcik commented on resource-aggregate/pb/commands.proto:
```diff
@@ -385,3 +385,96 @@ message ConfirmResourceCreateResponse {
+//     │ │  SignInResponse   │                               │ │                            │     
+//     │ │ <─────────────────│                               │ │                            │     
+//     └┬┘                   │                               └┬┘                            │     
+//      │                    │                               │ DeviceMetadataUpdatePending  │     
```
In this case it's DeviceMetadataUpdated as no confirmation is needed.

jkralik replied:
fixed

ondrejtomcik commented on resource-aggregate/pb/commands.proto:
```diff
@@ -385,3 +385,96 @@ message ConfirmResourceCreateResponse {
+    string device_id = 1;
+    string correlation_id = 2;
+    oneof confirm {
+        ShadowSynchronizationStatus shadow_synchronization_status = 3;
```
Why need to wrap it? If it's one of, it cannot be directly bool?

jkralik replied:
No. Maybe in future we want to add time how long it will be disabled ...

ondrejtomcik commented on resource-aggregate/pb/commands.proto:
```diff
@@ -385,3 +385,96 @@ message ConfirmResourceCreateResponse {
+// │OCF Server│          │Gateway│                         │Resource Aggregate│               │Event Bus│               │Client│
+// └──────────┘          └───────┘                         └──────────────────┘               └─────────┘               └──────┘
+message ShadowSynchronizationStatus {
+    bool disabled = 1;
```
I am for isEnabled :)

jkralik replied:
it about default value at proto it is `false` so if i changed it then all  must set it to true ... eg coap-gw but we don't want to do it ...

ondrejtomcik commented on resource-aggregate/pb/events.proto:
```diff
@@ -113,4 +113,32 @@ message ResourceStateSnapshotTaken {
+message DeviceMetadataUpdatePending {
+    string device_id = 1;
+    oneof update_pending {
+        ShadowSynchronizationStatus shadow_synchronization_status = 2;
```
Skip status, just ShadowSynchronization.

jkralik replied:
`shadow_synchronization_status` you want to skip it for all messages ?

jkralik replied:
fixed

ondrejtomcik commented on resource-aggregate/pb/commands.proto:
```diff
@@ -385,3 +385,96 @@ message ConfirmResourceCreateResponse {
     AuditContext audit_context = 100;
 }
 
+// Update Device Metadata - Connection status                           
```
Please create [here](https://github.com/plgd-dev/plgd.dev/tree/main/content/en/features) `device-shadow.md` and document what is it, how it works, how user can control it. (document feature itself including recent changes).

jkralik committed:
fix after CR
